lib/history/timeline.go

```diff
@@ -0,0 +1,234 @@
+	// RecordStatus records any changes that have occurred since the previous
+	// recorded status.
+	RecordStatus(status *pb.SystemStatus)
+	// GetEvents returns the currently stored list of events.
+	GetEvents() []*Event
+}
+
+// Cluster represents the overall status of a cluster.
+type Cluster struct {
+	// Status specifies the cluster status.
+	Status string
+	// Nodes specify the individual node statuses.
+	Nodes map[string]*Node
+}
+
+// diffCluster calculates the differences from the provided cluster and returns
+// the differences as a list of events.
+func (c *Cluster) diffCluster(cluster *Cluster) []*Event {
+	events := []*Event{}
+
+	// Compare cluster status
+	if c.Status != cluster.Status {
+		var event *Event
+		if cluster.Status == pb.SystemStatus_Running.String() {
+			event = NewClusterRecoveredEvent()
+		} else {
+			event = NewClusterDegradedEvent()
+		}
+		event.SetMetadata("old", c.Status)
+		event.SetMetadata("new", cluster.Status)
+		events = append(events, event)
+	}
+
+	// Keep track of removed nodes
+	removed := map[string]bool{}
+	for name := range c.Nodes {
+		removed[name] = true
+	}
+
+	// Nodes added or modified
+	for name, newNode := range cluster.Nodes {
+		if oldNode, ok := c.Nodes[name]; !ok {
+			event := NewNodeAddedEvent()
+			event.SetMetadata("node", name)
+			event.SetMetadata("new", newNode.Status)
+			events = append(events, event)
+
+			// Add new probes as well.
+			events = append(events, (&Node{}).diffNode(newNode)...)
+		} else {
+			events = append(events, oldNode.diffNode(newNode)...)
+			delete(removed, name)
+		}
+	}
+
+	// Nodes removed from the cluster
+	for name := range removed {
+		event := NewNodeRemovedEvent()
+		event.SetMetadata("node", name)
+		event.SetMetadata("old", c.Nodes[name].Status)
+		events = append(events, event)
+	}
+
+	return events
+}
+
+// Node represents the status of a node.
+type Node struct {
+	// Name specifies the name of the node.
+	Name string
+	// Status specifies the status of the node.
+	Status string
+	// MemberStatus specifies the node serf membership status.
+	MemberStatus string
+	// Probes specify the individual probe results.
+	Probes map[string]*Probe
+}
+
+// diffNode calculates the differences from the provided node and returns the
+// differences as a list of events.
+func (n *Node) diffNode(node *Node) []*Event {
+	events := []*Event{}
+
+	// Compare node status
+	if n.Status != node.Status {
+		var event *Event
+		if node.Status == pb.NodeStatus_Running.String() {
+			event = NewNodeRecoveredEvent()
+		} else {
+			event = NewNodeDegradedEvent()
+		}
+		event.SetMetadata("node", n.Name)
+		event.SetMetadata("old", n.Status)
+		event.SetMetadata("new", node.Status)
+		events = append(events, event)
+	}
+
+	// Keep track of removed probes
+	removed := map[string]bool{}
+	for name := range n.Probes {
+		removed[name] = true
+	}
+
+	// Probes added or modified
+	for name, newProbe := range node.Probes {
+		if oldProbe, ok := n.Probes[name]; !ok {
```

Comment:
Handle the happy (`ok == true`) case first and continue - w/o the need for `else`.